File: ProjectCS/UI/GroupPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SessionResultsCS.UI
{
    public class GroupPage : LoopPage
    {
        private Group _group;

        private MenuActionPicker _picker;


        public GroupPage(Group group)
        {
            _group = group;
            _picker = new MenuActionPicker([
                ("Перейти к студенту", GoToStudent),
                ("Изменить название группы", ChangeGroupName),
                ("Изменить курс группы", ChangeGroupCourse),
                ("Добавить студента", AddStudent),
                ("Удалить студента", DeleteStudent),
                ("Назад", StopLoop),
            ]);
        }

        protected override void Loop() 
        {
            Console.WriteLine($"==== {_group} ====");
            Console.WriteLine("Студенты:");
            PrintStudents();
            Console.WriteLine("Действия:");
            _picker.PickAndRun();
        }


        private void GoToStudent()
        {
            int selectedIndex = SelectStudent();
            if (selectedIndex != -1)
            {
                new StudentPage(_group.GetStudentAt(selectedIndex)).Start();
            }
        }

        private void ChangeGroupName()
        {
            try
            {
                Console.Write("Введите новое название группы: ");
                string name = Console.ReadLine()!;
                _group.Name = name;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Введено некорректное название");
            }
        }

        private void ChangeGroupCourse()
        {
            try
            {
                Console.Write("Введите новый курс группы: ");
                string course = Console.ReadLine()!;
                _group.Course = int.Parse(course);
            }
            catch (Exception e) when (e is ArgumentException || e is FormatException)
            {
                Console.WriteLine("Введен некорректный курс");
            }
        }

        private void AddStudent()
        {
            Console.Write("Введите номер студенческого билета нового студента: ");
            string studentID = Console.ReadLine()!;
            Console.Write("Введите имя нового студента: ");
            string firstName = Console.ReadLine()!;
            Console.Write("Введите фамилию нового студента: ");
            string lastName = Console.ReadLine()!;
            Console.Write("Введите отчество нового студента: ");
            string middleName = Console.ReadLine()!;

            try
            {
                Student student = new Student(_group, uint.Parse(studentID), firstName, lastName, middleName);
                _group.AddStudent(student);
            }
            catch (Exception e) when (e is ArgumentException || e is FormatException)
            {
                Console.WriteLine("Не удалось добавить студента");
            }
        }

        private void DeleteStudent()
        {
            int selectedIndex = SelectStudent();
            if (selectedIndex != -1)
            {
                _group.DeleteStudent(_group.GetStudentAt(selectedIndex));
            }
        }


        private void PrintStudents()
        {
            foreach (var (index, student) in _group.GetStudents().Index())
            {
                Console.WriteLine($"{index + 1}. {student}");
            }
        }

        private int SelectStudent()
        {
            Console.Write("Введите порядковый номер студента: ");
            string input = Console.ReadLine()!;
            if (int.TryParse(input, out int value) && (value >= 1 && value <= _group.GetStudentCount()))
            {
                return value - 1;
            }
            else
            {
                Console.WriteLine("Введён неверный номер");
                return -1;
            }
        }
    }
}

File: Project/UI/GroupPage.cpp
#include "UI/GroupPage.hpp"

#include "Data/Speciality.hpp"

#include <format>
#include <iostream>


GroupPage::GroupPage(std::shared_ptr<Group> group) {
	m_group = group;
	SetPicker(std::make_unique<MenuActionPicker>(std::vector<MenuActionPicker::Item>{
		{ "Перейти к студенту", [this]() { GoToStudent(); } },
		{ "Изменить название группы", [this]() { ChangeGroupName(); } },
		{ "Изменить курс группы", [this]() { ChangeGroupCourse(); } },
		{ "Добавить студента", [this]() { AddStudent(); } },
		{ "Удалить студента", [this]() { DeleteStudent(); } },
		{ "Назад", [this]() { StopLoop(); } }
	}));
}

void GroupPage::Loop() {
	std::cout << std::format("==== {} ====\n", m_group->ToString());
	std::cout << "Студенты:\n";
	PrintStudents();
}

void GroupPage::GoToStudent() {
	int selectedIndex = SelectStudent();
	if (selectedIndex != -1) {
		StudentPage(m_group->GetStudentAt(selectedIndex)).Start();
	}
}

void GroupPage::ChangeGroupName() {
	try {
		std::cout << "Введите новое название группы: ";
		std::string name;
		std::getline(std::cin, name);
		m_group->SetName(name);
	} catch (std::invalid_argument) {
		std::cout << "Введено некорректное название\n";
	}
}

void GroupPage::ChangeGroupCourse() {
	try {
		std::cout << "Введите новый курс группы: ";
		std::string input;
		std::getline(std::cin, input);
		int course = std::stoi(input);
		m_group->SetCourse(course);
	} catch (std::invalid_argument) {
		std::cout << "Введен некорректный курс\n";
	} catch (std::out_of_range) {
		std::cout << "Введен некорректный курс\n";
	}
}

void GroupPage::AddStudent() {
	std::cout << "Введите номер студенческого билета нового студента: ";
	std::string studentIDInput;
	std::getline(std::cin, studentIDInput);

	std::cout << "Введите имя нового студента: ";
	std::string firstName;
	std::getline(std::cin, firstName);

	std::cout << "Введите фамилию нового студента: ";
	std::string lastName;
	std::getline(std::cin, lastName);

	std::cout << "Введите отчество нового студента: ";
	std::string middleName;
	std::getline(std::cin, middleName);

	try {
		unsigned int studentID = std::stoul(studentIDInput);
		m_group->AddStudent(std::make_shared<Student>(studentID, firstName, lastName, middleName));
	} catch (std::invalid_argument) {
		std::cout << "Не удалось добавить студента\n";
	} catch (std::out_of_range) {
		std::cout << "Не удалось добавить студента\n";
	}
}

void GroupPage::DeleteStudent() {
	int selectedIndex = SelectStudent();
	if (selectedIndex != -1) {
		m_group->DeleteStudent(m_group->GetStudentAt(selectedIndex));
	}
}

void GroupPage::PrintStudents() {
	for (int i = 0; i < m_group->GetStudentCount(); i++) {
		std::cout << std::format("{}. {}\n", i + 1, m_group->GetStudentAt(i)->ToString());
	}
}

int GroupPage::SelectStudent() {
	int value;
	std::cout << "Введите порядковый номер студента: ";
	if ((std::cin >> value) && (value >= 1 && value <= m_group->GetStudentCount())) {
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		return value - 1;
	} else {
		std::cin.clear();
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		std::cout << "Введён неверный номер\n";
		return -1;
	}
}